### ISSUE TITLE
ClassNotFoundException when trying to import Spring Boot jar

### ISSUE THREAD

**Issue by MarkusJx** [MarkusJx/node-java-bridge#44]
Hello @MarkusJx 

the lib path was only for DLLs which a printer need

same problems / error as before
and the "missing" class from the error is in spring-core-4.3.3.RELEASE.jar available
so no idea whats wrong

Code
[main.txt](https://github.com/MarkusJx/node-java-bridge/files/10086409/main.txt)
Output
[output.txt](https://github.com/MarkusJx/node-java-bridge/files/10086406/output.txt)
Error
[error.txt](https://github.com/MarkusJx/node-java-bridge/files/10086405/error.txt)

@MarkusJx maybe you could also try something with spring franework
for your information
my project is used as normal spring boot and also in this case like IoC Container 
https://www.geeksforgeeks.org/spring-ioc-container/
https://www.baeldung.com/inversion-control-and-dependency-injection-in-spring

_Originally posted by @tripower in https://github.com/MarkusJx/node-java-bridge/discussions/37#discussioncomment-4228375_

**Comment by MarkusJx** [comment]
Spring boot uses its own class loader, thus, the class path must be set *before* jvm startup. See this comment: https://github.com/MarkusJx/node-java-bridge/discussions/37#discussioncomment-4277361

**Comment by tripower** [comment]
ok
so that should work?

...
for (const file of arFiles) {
    java.classpath.append(file);
    console.log("append file to cp: " + file);
}

java.ensureJvm({
    opts: [
        '-Djava.library.path=${localservicejarsPath};',
        '-Dspring.application.admin.enabled=true;'
    ]
});
...

**Comment by tripower** [comment]
should we make a extra discussion for Spring Boot or 
do you extend your readme / help
that other will also find the specific information?

greetings
Klaus

**Comment by MarkusJx** [comment]
No, that won't work. `java.classpath.append` itself calls `ensureJvm` with no arguments, thus, rendering your own call to `ensureJvm` useless, as it will do nothing.

This is also mentioned in the [docs for `ensureJvm`](https://markusjx.github.io/node-java-bridge/functions/ensureJvm.html):
> This method is also called every time with no arguments when any call to the jvm is done in another method.

#49 adds two things, which will make this whole process easier:
1. `ensureJvm` now returns `true` if the jvm has been created, `false` if it has already been created and the call to `ensureJvm` was a no-op
2. There is now an option `classpath`, which allows you to add your dependencies as an array. It even parses your `-Djava.library.path=` startup option (if present) and adds this into one option, replacing the separators with the appropriate ones for your os along the way. The elements of the `classpath` array won't be altered though, so adding multiple classpath elements into one big string is not recommended.

The whole point of this PR is to make altering the startup classpath more convenient, first, you are notified if the classpath can't be altered anymore and second, it's now easier to alter the startup classpath and it should be more clear whether you're setting the startup or runtime classpath, as the `classpath` option of `ensureJvm` alters the *startup* classpath and `classpath.append` or `appendClasspath` alters the *runtime* classpath, which may not result in an altered Spring Boot classpath.

That being said, the correct way of setting your application's classpath would be
```ts
for (const file of arFiles) {
	console.log("append file to cp: " + file);
}

// ensureJvm should be the first call to java-bridge
java.ensureJvm({
	opts: [
		`-Djava.library.path=${argFiles.join(';')};`,
		'-Dspring.application.admin.enabled=true;'
	]
});
```

After version `2.1.8` has been released, you should be able to do something like this:
```ts
const started = java.ensureJvm({
	opts: [
		'-Dspring.application.admin.enabled=true;'
	],
	classpath: argFiles,
});

if (!started) {
	throw new Error('The jvm has already been started - Failed to set startup classpath!');
}
```

**Comment by MarkusJx** [comment]
> should we make a extra discussion for Spring Boot or do you extend your readme / help that other will also find the specific information?
> 
> greetings Klaus

Others may find the information in this issue, so just leaving this information inside this issue will work.
I've also extended the documentation of `ensureJvm` adding some information about altering the classpath.

Basically, everyone who knows how to use google should be able to find anything they need :)

**Comment by tripower** [comment]
ok
thx for the perfect info
wozld it also be possible to set a path instead of jars at the opts?

but I have to wait for the new version

thx in advance
Klaus

**Comment by tripower** [comment]
cool😍
sounds good 🤗

**Comment by MarkusJx** [comment]
You can try out the latest changes in the `2.1.8-beta.1` release

**Comment by tripower** [comment]
ok
thx

**Comment by MarkusJx** [comment]
ok, well, this is weird, the [documentation on the class path](https://docs.oracle.com/javase/7/docs/technotes/tools/windows/classpath.html) clearly says that wildcard imports should work:
> Class path entries can contain the basename wildcard character \*, which is considered equivalent to specifying a list of all the files in the directory with the extension .jar or .JAR. For example, the class path entry foo/* specifies all JAR files in the directory named foo. A classpath entry consisting simply of \* expands to a list of all the jar files in the current directory.
>
> A class path entry that contains * will not match class files. To match both classes and JAR files in a single directory foo, use either foo;foo/* or foo/*;foo. The order chosen determines whether the classes and resources in foo are loaded before JAR files in foo, or vice versa.
>
> Subdirectories are not searched recursively. For example, foo/* looks for JAR files only in foo, not in foo/bar, foo/baz, etc.

But if that doesn't work, just use glob, that's anyway the better alternative as it should give you more control regarding which files you want to import

**Comment by tripower** [comment]
it has worked at joeferners bridge in this way
have you try it on Windows?
for me not a great problem, only for feedback

if have tried 
slash, backslash, double 
absolute, relative 
...

greetings
Klaus

**Comment by MarkusJx** [comment]
Ok, tested it myself, doesn't work, great

**Comment by tripower** [comment]
😂👍

**Comment by MarkusJx** [comment]
Yes, I did look at that, so I've added an option to both `appendClasspath` and the `classpath` argument to match files using glob patterns. Directories will no longer be automatically resolved, only using glob patterns will work from now on:
```ts
// Import a single jar
appendClasspath('/path/to/your/file.jar');

// Import multiple jars from a directory
appendClasspath('/path/to/*.jar');
```

**Comment by tripower** [comment]
ok
was only surprised that there worked directories
but for me this solution is ok
happy that so much is working now 🥳

**Comment by MarkusJx** [comment]
Well, not everything is working right now, adding glob support for some reason broke the package on win32. So I'll fix that in the near future.

napi-rs on win32 is...weird

**Comment by tripower** [comment]
sounds real tough
and could be a fulltime job with this
mass on needed features and "bugs"

greetings Klaus

**Comment by MarkusJx** [comment]
It's working now. For some reason. I've created a new release for you to test this: `2.1.8-beta.2`

So just let me know if this works for you and if you have any other issues

**Comment by tripower** [comment]
ok
at the moment I have no other wishes 
but who knows whats coming next 😂🙏

**Comment by tripower** [comment]
@MarkusJx 
have now made a short test with beta 2 
looks good and all needed java objects can be instantiated
and getmethods of the objects shows all I need
further tests with real printer will follow
hope that works with DLLs in the classpath
but that will take some time 
so you can close this issue and make a version ?

thx for the fixes so far!!!

greetings 
Klaus

**Comment by MarkusJx** [comment]
ok, great, will create a new release once #49 is merged

**Comment by tripower** [comment]
thx, perfect

**Comment by MarkusJx** [comment]
The changes are now live in the latest release, `v2.2.0` 🎉

**Comment by tripower** [comment]
yippie 
party time 😂🙏🎉🥳
thx
